gammapy/cube/psf_kernel.py

```diff
@@ -0,0 +1,310 @@
+
+    # loop over images and fill map
+    for img, idx in kernel_map.iter_by_image():
+        img += vals.value
+
+    # downsample the psf kernel map. Take the average
+    kernel_map = kernel_map.downsample(factor, preserve_counts=False)
+
+    if normalize:
+        # normalize each image in map
+        for img, idx in kernel_map.iter_by_image():
+            norm = np.sum(img)
+            img /= norm
+
+    return kernel_map
+
+
+def energy_dependent_table_psf_to_kernel_map(table_psf, geom, normalize=True, factor=4):
+    """Compute an energy dependent PSF kernel on a given MapGeom.
+
+    The PSF is estimated by oversampling defined by a given factor.
+
+    Parameters
+    ----------
+    table_psf : `~gammapy.irf.EnergyDependentTablePSF`
+        the input table PSF
+    geom : `~gammapy.maps.MapGeom`
+        the target geometry. The PSF kernel will be centered on the spatial centre.
+        the geometry axes should contain an energy MapAxis.
+    normalize : bool
+        normalize the PSF kernel (per energy)
+    factor : int
+        the oversample factor to compute the PSF
+    """
+    # Find energy axis in geom
+    energy_idx = -1
+    for i, axis in enumerate(geom.axes):
+        if axis.type is 'energy':
+            energy_idx = i
+
+    if energy_idx == -1:
+        raise ValueError("No energy axis in target geometry for PSFKernel")
+
+    energy_unit = u.Unit(geom.axes[energy_idx].unit)
+
+    # TODO: change the logic to support non-regular geometry. This would allow energy dependent sizes for the kernel.
+    if geom.is_regular is False:
+        raise ValueError("Non regular geometries non supported yet.")
+
+    # First upsample spatial geom
+    upsampled_image_geom = geom.to_image().upsample(factor)
+
+    # get center coordinate
+    center_coord = upsampled_image_geom.center_coord * u.deg
+
+    # get coordinates
+    map_c = upsampled_image_geom.get_coord()
+
+    # get solid angles
+    solid_angles = upsampled_image_geom.solid_angle()
+
+    # compute distances to map center
+    rads = angular_separation(center_coord[0], center_coord[1], map_c.lon * u.deg, map_c.lat * u.deg)
+
+    # Create map
+    kernel_map = Map.from_geom(geom=upsampled_image_geom.to_cube(axes=geom.axes),
+                               unit='')
+
+    # loop over images
+    for img, idx in kernel_map.iter_by_image():
+        energy = geom.axes[energy_idx].center[idx[energy_idx]] * energy_unit
+        vals = table_psf.evaluate(energy=energy, rad=rads).reshape(img.shape) * solid_angles
+        img += vals.value
+
+    # downsample the psf kernel map. Take the average
+    kernel_map = kernel_map.downsample(factor, preserve_counts=False)
+
+    if normalize:
+        # normalize each image in map
+        for img, idx in kernel_map.iter_by_image():
+            norm = np.sum(img)
+            img /= norm
+
+    return kernel_map
+
+
+class PSFKernel(object):
+    """PSF kernel for `~gammapy.maps.Map`.
```

Comment:
Does this class only support 2D and 3D with energy axis?
Or are extra axes also tolerated and work to some extent?
Suggest to clarify in the docstring (but not spend any time to make extra axes work if they don't already).

Reply:
Extra axes are also tolerated but filled the same way.
Docstring clarified.